File: leetcode/serialize_and_deserialize_binary_tree/serialize_and_deserialize_binary_tree.java
package serialize_and_deserialize_binary_tree;

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "n";

        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            TreeNode curr = q.poll();

            if (curr == null) {
                sb.append("n ");
                continue;
            }

            sb.append(curr.val).append(" ");
            q.add(curr.left);
            q.add(curr.right);
        }

        sb.setLength(sb.length() - 1);
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data == null || data.equals("n")) return null;

        String[] nodes = data.split(" ");
        int i = 0;

        TreeNode root = new TreeNode(Integer.parseInt(nodes[i++]));
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            TreeNode curr = q.poll();

            if (i < nodes.length && !nodes[i].equals("n")) {
                curr.left = new TreeNode(Integer.parseInt(nodes[i]));
                q.add(curr.left);
            }
            i++;

            if (i < nodes.length && !nodes[i].equals("n")) {
                curr.right = new TreeNode(Integer.parseInt(nodes[i]));
                q.add(curr.right);
            }
            i++;
        }

        return root;
    }
}

File: leetcode/serialize_and_deserialize_binary_tree/serialize_and_deserialize_binary_tree.cpp
/*
 * Time Complexity:
 *   - serialize: O(N)
 *     - Performs a level-order traversal, visiting each node once.
 *   - deserialize: O(N)
 *     - Parses each value in the string once and reconstructs the tree level by level.
 *   - N is the number of nodes in the tree.
 *
 * Space Complexity:
 *   - O(N)
 *     - For the queue used in both serialization and deserialization.
 *     - The output string in `serialize` also takes O(N) space.
 */

#include <bits/stdc++.h>
#include <iostream>
#include <string>
#include <queue>

using std::string;
using std::queue;
using std::to_string;
using std::stoi;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        string s = "";
        TreeNode* currNode;

        while(!q.empty()) {
            currNode = q.front();
            q.pop();

            if(!currNode) {
                s += "n ";
                continue;
            } 

            s += to_string(currNode->val) + ' ';
            q.push(currNode->left);
            q.push(currNode->right);
        }
        s.pop_back();
        return s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data.empty() || data == "n") return nullptr;

        int l = 0, r = 0, number;
        string numberStr = "";

        while(r < data.size() && data[r] != ' ') r++;

        numberStr = data.substr(l, r-l);
        number = stoi(numberStr);

        r++;
        l=r;

        TreeNode* root = new TreeNode(number);
        TreeNode* currNode;

        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()) {
            currNode = q.front();
            q.pop();

            while(r < data.size() && data[r] != ' ') r++;

            numberStr = data.substr(l, r-l);
            if(numberStr != "n") {
                number = stoi(numberStr);
                TreeNode* left = new TreeNode(number);
                currNode->left = left;
                q.push(left);
            }

            r++;
            l=r;

            while(r < data.size() && data[r] != ' ') r++;

            numberStr = data.substr(l, r-l);
            if(numberStr != "n") {
                number = stoi(numberStr);
                TreeNode* right = new TreeNode(number);
                currNode->right = right;
                q.push(right);
            }

            r++;
            l=r;
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));